packages/queued-request-controller/src/QueuedRequestController.ts

```diff
@@ -168,61 +245,87 @@ export class QueuedRequestController extends BaseController<
 
     await this.messagingSystem.call(
       'NetworkController:setActiveNetwork',
-      request.networkClientId,
+      originNetworkClientId,
     );
   }
 
-  #updateCount(change: -1 | 1) {
+  /**
+   * Update the queued request count.
+   */
+  #updateCount() {
     this.update((state) => {
-      state.queuedRequestCount += change;
+      state.queuedRequestCount = this.#requestQueue.length;
     });
   }
 
   /**
-   * Enqueues a new request for sequential processing in the request queue. This function manages the order of
-   * requests, ensuring they are executed one after the other to prevent concurrency issues and maintain proper
-   * execution flow.
+   * Enqueue a request to be processed in a batch with other requests from the same origin.
+   *
+   * We process requests one origin at a time, so that requests from different origins do not get
+   * interwoven, and so that we can ensure that the globally selected network matches the dapp-
+   * selected network. Request are executed in exactly the same order they come in.
    *
    * @param request - The JSON-RPC request to process.
-   * @param requestNext - A function representing the next steps for processing this request. It returns a promise that
-   * resolves when the request is complete.
-   * @returns A promise that resolves when the enqueued request and any subsequent asynchronous
-   * operations are fully processed. This allows you to await the completion of the enqueued request before continuing
-   * with additional actions. If there are multiple enqueued requests, this function ensures they are processed in
-   * the order they were enqueued, guaranteeing sequential execution.
+   * @param requestNext - A function representing the next steps for processing this request.
+   * @returns A promise that resolves when the given request has been fully processed.
    */
   async enqueueRequest(
     request: QueuedRequestMiddlewareJsonRpcRequest,
     requestNext: () => Promise<void>,
-  ) {
-    this.#updateCount(1);
-    if (this.state.queuedRequestCount > 1) {
-      try {
-        await this.currentRequest;
-      } catch (_error) {
-        // error ignored - this is handled in the middleware instead
-        this.#updateCount(-1);
-      }
+  ): Promise<void> {
+    if (this.#originOfCurrentBatch === undefined) {
+      this.#originOfCurrentBatch = request.origin;
     }
 
-    const processCurrentRequest = async () => {
-      try {
-        if (
-          request.method !== 'wallet_switchEthereumChain' &&
-          request.method !== 'wallet_addEthereumChain'
-        ) {
-          await this.#switchNetworkIfNecessary(request);
-        }
+    try {
+      // Queue request for later processing
+      // Network switch is handled when this batch is processed
+      if (
+        this.state.queuedRequestCount > 0 ||
+        this.#originOfCurrentBatch !== request.origin
```

Comment:
if the `queuedRequestCount` is zero is it possible for the `this.#originOfCurrentBatch !== request.origin` ? I would assume that if queuedRequestCount was zero then `this.#originOfCurrentBatch === undefined` condition would have also been true just above and so `this.#originOfCurrentBatch` would have been  set to equal `request.origin`

Reply:
Yes, it is possible if the current batch has another batch queued after it.

For example, lets say we get requests from these sites in this order:

1. Test dapp
2. Polygon
3. Test dapp

For the third request, we'd see that the origin matches the current batch origin. But we still need to queue it to preserve request order

Reply:
Yes, it is possible. Note that this is the queued request count, e.g. the number of requests queued up, not currently processing. There may be any number of requests currently processing.

e.g. we might have requests from these origins come in, in this order:

1. Test dapp
2. Test dapp

Or in this scenario:

1. Test dapp
2. Uniswap


In both examples, the queued request count would be zero when the second request was processed, but the batch origin would differ.